diff --git a/common/base.c b/common/base.c
index 0e882e4b..4139d42a 100644
--- a/common/base.c
+++ b/common/base.c
@@ -429,6 +429,7 @@ REALIGN_STACK void x264_param_default( x264_param_t *param )
     param->pf_log = x264_log_default;
     param->p_log_private = NULL;
     param->i_log_level = X264_LOG_INFO;
+    param->i_log_file_level = X264_LOG_INFO;
 
     /* */
     param->analyse.intra = X264_ANALYSE_I4x4 | X264_ANALYSE_I8x8;
@@ -1202,6 +1203,13 @@ REALIGN_STACK int x264_param_parse( x264_param_t *p, const char *name, const cha
     }
     OPT("log")
         p->i_log_level = atoi(value);
+    OPT("log-file")
+        p->psz_log_file = strdup(value);
+    OPT("log-file-level")
+        if( !parse_enum( value, x264_log_file_level_names, &p->i_log_file_level ) )
+            p->i_log_file_level += X264_LOG_NONE;
+        else
+            p->i_log_file_level = atoi(value);
     OPT("dump-yuv")
         CHECKED_ERROR_PARAM_STRDUP( p->psz_dump_yuv, p, value );
     OPT2("analyse", "partitions")
diff --git a/common/common.c b/common/common.c
index 1d0dee72..4aca3530 100644
--- a/common/common.c
+++ b/common/common.c
@@ -26,6 +26,8 @@
 
 #include "common.h"
 
+static void x264_log_file( char *, int, const char *, va_list );
+
 /****************************************************************************
  * x264_log:
  ****************************************************************************/
@@ -41,4 +43,42 @@ void x264_log( x264_t *h, int i_level, const char *psz_fmt, ... )
             h->param.pf_log( h->param.p_log_private, i_level, psz_fmt, arg );
         va_end( arg );
     }
+
+    if( h && h->param.psz_log_file && i_level <= h->param.i_log_file_level )
+    {
+        va_list arg;
+        va_start( arg, psz_fmt );
+        x264_log_file( h->param.psz_log_file, i_level, psz_fmt, arg );
+        va_end( arg );
+    }
+}
+
+static void x264_log_file( char *p_file_name, int i_level, const char *psz_fmt, va_list arg )
+{
+    char *psz_prefix;
+    switch( i_level )
+    {
+        case X264_LOG_ERROR:
+            psz_prefix = "error";
+            break;
+        case X264_LOG_WARNING:
+            psz_prefix = "warning";
+            break;
+        case X264_LOG_INFO:
+            psz_prefix = "info";
+            break;
+        case X264_LOG_DEBUG:
+            psz_prefix = "debug";
+            break;
+        default:
+            psz_prefix = "unknown";
+            break;
+    }
+    FILE *p_log_file = x264_fopen( p_file_name, "ab" );
+    if( p_log_file )
+    {
+        fprintf( p_log_file, "x264 [%s]: ", psz_prefix );
+        vfprintf( p_log_file, psz_fmt, arg );
+        fclose( p_log_file );
+    }
 }
diff --git a/encoder/encoder.c b/encoder/encoder.c
index 14cb9a91..ac34715b 100644
--- a/encoder/encoder.c
+++ b/encoder/encoder.c
@@ -1125,7 +1125,7 @@ static int validate_parameters( x264_t *h, int b_open )
     if( h->param.rc.f_aq_strength == 0 )
         h->param.rc.i_aq_mode = 0;
 
-    if( h->param.i_log_level < X264_LOG_INFO )
+    if( h->param.i_log_level < X264_LOG_INFO && (!h->param.psz_log_file || h->param.i_log_file_level < X264_LOG_INFO) )
     {
         h->param.analyse.b_psnr = 0;
         h->param.analyse.b_ssim = 0;
@@ -2938,7 +2938,7 @@ cont:
 
         int b_intra = IS_INTRA( h->mb.i_type );
         int b_skip = IS_SKIP( h->mb.i_type );
-        if( h->param.i_log_level >= X264_LOG_INFO || h->param.rc.b_stat_write )
+        if( h->param.i_log_level >= X264_LOG_INFO || (h->param.psz_log_file && h->param.i_log_file_level >= X264_LOG_INFO) || h->param.rc.b_stat_write )
         {
             if( !b_intra && !b_skip && !IS_DIRECT( h->mb.i_type ) )
             {
@@ -2958,7 +2958,7 @@ cont:
             }
         }
 
-        if( h->param.i_log_level >= X264_LOG_INFO )
+        if( h->param.i_log_level >= X264_LOG_INFO || (h->param.psz_log_file && h->param.i_log_file_level >= X264_LOG_INFO) )
         {
             if( h->mb.i_cbp_luma | h->mb.i_cbp_chroma )
             {
diff --git a/x264.c b/x264.c
index 7b016871..1be09093 100644
--- a/x264.c
+++ b/x264.c
@@ -262,9 +262,32 @@ static int  parse( int argc, char **argv, x264_param_t *param, cli_opt_t *opt );
 static int  encode( x264_param_t *param, cli_opt_t *opt );
 
 /* logging and printing for within the cli system */
+static char *psz_log_file       = NULL;
+static int   cli_log_file_level = -1;
+
+static inline void x264_log_done()
+{
+    if( psz_log_file ) free( psz_log_file );
+    psz_log_file = NULL;
+}
+
+static inline void x264_log_init( const char *file_name )
+{
+    x264_log_done();
+    psz_log_file = strdup( file_name );
+}
+
 static int cli_log_level = X264_LOG_INFO;
 void x264_cli_log( const char *name, int i_level, const char *fmt, ... )
 {
+    if( psz_log_file && *psz_log_file && i_level <= cli_log_file_level )
+    {
+        va_list arg;
+        va_start( arg, fmt );
+        x264_cli_log_file( psz_log_file, i_level, fmt, arg );
+        va_end( arg );
+    }
+
     if( i_level > cli_log_level )
         return;
     char *s_level;
@@ -293,8 +316,46 @@ void x264_cli_log( const char *name, int i_level, const char *fmt, ... )
     va_end( arg );
 }
 
+void x264_cli_log_file( char *p_file_name, int i_level, const char *psz_fmt, va_list arg )
+{
+    char *psz_prefix;
+    switch( i_level )
+    {
+        case X264_LOG_ERROR:
+            psz_prefix = "error";
+            break;
+        case X264_LOG_WARNING:
+            psz_prefix = "warning";
+            break;
+        case X264_LOG_INFO:
+            psz_prefix = "info";
+            break;
+        case X264_LOG_DEBUG:
+            psz_prefix = "debug";
+            break;
+        default:
+            psz_prefix = "unknown";
+            break;
+    }
+    FILE *p_log_file = x264_fopen( p_file_name, "ab" );
+    if( p_log_file )
+    {
+        fprintf( p_log_file, "x264 [%s]: ", psz_prefix );
+        vfprintf( p_log_file, psz_fmt, arg );
+        fclose( p_log_file );
+    }
+ }
+
 void x264_cli_printf( int i_level, const char *fmt, ... )
 {
+    if( psz_log_file && *psz_log_file )
+    {
+        va_list arg;
+        va_start( arg, fmt );
+        x264_cli_log_file( psz_log_file, X264_LOG_INFO, fmt, arg );
+        va_end( arg );
+    }
+
     if( i_level > cli_log_level )
         return;
     va_list arg;
@@ -409,6 +470,7 @@ REALIGN_STACK int main( int argc, char **argv )
     free( argv );
 #endif
 
+    x264_log_done();
     return ret;
 }
 
@@ -930,8 +992,12 @@ static void help( x264_param_t *defaults, int longhelp )
     H1( "      --no-progress           Don't show the progress indicator while encoding\n" );
     H0( "      --quiet                 Quiet Mode\n" );
     H1( "      --log-level <string>    Specify the maximum level of logging [\"%s\"]\n"
-        "                                  - %s\n", strtable_lookup( x264_log_level_names, cli_log_level - X264_LOG_NONE ),
-                                       stringify_names( buf, x264_log_level_names ) );
+        "                                  - %s\n", strtable_lookup( x264_log_file_level_names, cli_log_level - X264_LOG_NONE ),
+                                       stringify_names( buf, x264_log_file_level_names ) );
+    H1( "      --log-file <string>     Save log to file\n" );
+    H1( "      --log-file-level <int>  Log-file level information [\"%s\"]\n"
+        "                                  - %s\n", strtable_lookup( x264_log_file_level_names, defaults->i_log_file_level - X264_LOG_NONE ),
+                                       stringify_names( buf, x264_log_file_level_names ) );
     H1( "      --psnr                  Enable PSNR computation\n" );
     H1( "      --ssim                  Enable SSIM computation\n" );
     H1( "      --threads <integer>     Force a specific number of threads\n" );
@@ -994,6 +1060,8 @@ typedef enum
     OPT_TIMEBASE,
     OPT_PULLDOWN,
     OPT_LOG_LEVEL,
+    OPT_LOG_FILE,
+    OPT_LOG_FILE_LEVEL,
     OPT_VIDEO_FILTER,
     OPT_INPUT_FMT,
     OPT_INPUT_RES,
@@ -1130,6 +1198,8 @@ static struct option long_options[] =
     { "quiet",             no_argument, NULL, OPT_QUIET },
     { "verbose",           no_argument, NULL, 'v' },
     { "log-level",   required_argument, NULL, OPT_LOG_LEVEL },
+    { "log-file",          required_argument, NULL, OPT_LOG_FILE },
+    { "log-file-level",    required_argument, NULL, OPT_LOG_FILE_LEVEL },
     { "no-progress",       no_argument, NULL, OPT_NOPROGRESS },
     { "dump-yuv",    required_argument, NULL, 0 },
     { "sps-id",      required_argument, NULL, 0 },
@@ -1423,6 +1493,7 @@ static int parse( int argc, char **argv, x264_param_t *param, cli_opt_t *opt )
 
     x264_param_default( &defaults );
     cli_log_level = defaults.i_log_level;
+    cli_log_file_level = defaults.i_log_file_level;
 
     memset( &input_opt, 0, sizeof(cli_input_opt_t) );
     memset( &output_opt, 0, sizeof(cli_output_opt_t) );
@@ -1496,12 +1567,21 @@ static int parse( int argc, char **argv, x264_param_t *param, cli_opt_t *opt )
                 cli_log_level = param->i_log_level = X264_LOG_DEBUG;
                 break;
             case OPT_LOG_LEVEL:
-                if( !parse_enum_value( optarg, x264_log_level_names, &cli_log_level ) )
+                if( !parse_enum_value( optarg, x264_log_file_level_names, &cli_log_level ) )
                     cli_log_level += X264_LOG_NONE;
                 else
                     cli_log_level = atoi( optarg );
                 param->i_log_level = cli_log_level;
                 break;
+            case OPT_LOG_FILE:
+                x264_log_init( optarg );
+                goto generic_option;
+            case OPT_LOG_FILE_LEVEL:
+                if( !parse_enum_value( optarg, x264_log_file_level_names, &cli_log_file_level ) )
+                    cli_log_file_level += X264_LOG_NONE;
+                else
+                    cli_log_file_level = atoi( optarg );
+                goto generic_option;
             case OPT_NOPROGRESS:
                 opt->b_progress = 0;
                 break;
@@ -1683,7 +1763,7 @@ generic_option:
     {
         if( thread_input->open_file( NULL, &opt->hin, &info, NULL ) )
         {
-            fprintf( stderr, "x264 [error]: threaded input failed\n" );
+            x264_cli_log( "x264", X264_LOG_ERROR, "threaded input failed\n" );
             return -1;
         }
         cli_input = *thread_input;
@@ -2089,7 +2169,7 @@ fail:
     fprintf( stderr, "\n" );
 
     if( b_ctrl_c )
-        fprintf( stderr, "aborted at input frame %d, output frame %d\n", opt->i_seek + i_frame, i_frame_output );
+        x264_cli_printf( X264_LOG_INFO, "aborted at input frame %d, output frame %d\n", opt->i_seek + i_frame, i_frame_output );
 
     cli_output.close_file( opt->hout, largest_pts, second_largest_pts );
     opt->hout = NULL;
@@ -2100,7 +2180,7 @@ fail:
                      (double)( i_end - i_start );
         int secs = (i_end - i_start) / 1000000;
 
-        fprintf( stderr, "encoded %d frames, %.2f fps, %.2f kb/s, duration %d:%02d:%02d.%02d\n", i_frame_output, fps,
+        x264_cli_printf( X264_LOG_INFO, "encoded %d frames, %.2f fps, %.2f kb/s, duration %d:%02d:%02d.%02d\n", i_frame_output, fps,
                  (double) i_file * 8 / ( 1000 * duration ),
                  secs/3600, (secs/60)%60, secs%60, (int)((i_end - i_start)%1000000/10000) );
     }
diff --git a/x264.h b/x264.h
index a79decf0..0c1b58f2 100644
--- a/x264.h
+++ b/x264.h
@@ -237,6 +237,7 @@ static const char * const x264_transfer_names[] = { "", "bt709", "undef", "", "b
 static const char * const x264_colmatrix_names[] = { "GBR", "bt709", "undef", "", "fcc", "bt470bg", "smpte170m", "smpte240m", "YCgCo", "bt2020nc", "bt2020c",
                                                      "smpte2085", "chroma-derived-nc", "chroma-derived-c", "ICtCp", 0 };
 static const char * const x264_nal_hrd_names[] = { "none", "vbr", "cbr", 0 };
+static const char * const x264_log_file_level_names[] = { "none", "error", "warning", "info", "debug", 0 };
 static const char * const x264_avcintra_flavor_names[] = { "panasonic", "sony", 0 };
 
 /* Colorspace type */
@@ -389,6 +390,8 @@ typedef struct x264_param_t
     void        (*pf_log)( void *, int i_level, const char *psz, va_list );
     void        *p_log_private;
     int         i_log_level;
+    int         i_log_file_level;
+    char        *psz_log_file;  /* filename (in UTF-8) of log-file */
     int         b_full_recon;   /* fully reconstruct frames, even when not necessary for encoding.  Implied by psz_dump_yuv */
     char        *psz_dump_yuv;  /* filename (in UTF-8) for reconstructed frames */
 
diff --git a/x264cli.h b/x264cli.h
index a7a6ad17..fc55a2ec 100644
--- a/x264cli.h
+++ b/x264cli.h
@@ -74,6 +74,7 @@ static inline char *get_filename_extension( char *filename )
 }
 
 void x264_cli_log( const char *name, int i_level, const char *fmt, ... );
+void x264_cli_log_file( char *p_file_name, int i_level, const char *psz_fmt, va_list arg );
 void x264_cli_printf( int i_level, const char *fmt, ... );
 int x264_cli_autocomplete( const char *prev, const char *cur );
 
